### PR TITLE
Stop enforcing W504; Remove check_flake8.sh, use flake8 directly

### LINKED ISSUE #758: Remove check_flake8.sh
This was created to help us check new code for flake8 compliance, even though the codebase at the time was really bad.

We now have brought the codebase up-to-date, and check_flake8 is no longer needed.

This can be done as soon as #409 is closed.

It will require:
* Switching .pre-commit-config.yaml to use the default flake8 check
* Verifying that flake8 still catches problems in the CI (in test, examples, src, and root)
* Verifying that flake8 still catches missing documentation in the CI for new code

### PR DESCRIPTION
**Pull request by gitanshu**
Closes #758

### CODE REVIEW

File: src/garage/envs/base.py [hunk 1]
```diff
@@ -94,9 +93,11 @@ def _close_viewer_window(self):
                     glfw.destroy_window(self.env.viewer.window)
             elif any(package in self.env.spec._entry_point
                      for package in KNOWN_GYM_NOT_CLOSE_VIEWER):
-                if (hasattr(self.env, 'viewer') and
-                    (isinstance(self.env.viewer, Viewer)
-                     or isinstance(self.env.viewer, SimpleImageViewer))):
+                # yapf: disable
```
Comment (ryanjulian) [inline src/garage/envs/base.py]: hmm -- if this version of YAPF still making the wrong splits?

File: src/garage/envs/base.py [hunk 2]
```diff
@@ -94,9 +93,11 @@ def _close_viewer_window(self):
...
+                if (hasattr(self.env, 'viewer')
+                    and (isinstance(self.env.viewer, Viewer)
+                         or isinstance(self.env.viewer, SimpleImageViewer))):
+                    # yapf: enable
```
Comment (ryanjulian) [inline src/garage/envs/base.py]: is YAPF not behaving?

File: src/garage/tf/algos/dqn.py [hunk 3]
```diff
@@ -185,8 +184,10 @@ def train_once(self, itr, paths):
         self.episode_rewards.extend(paths['undiscounted_returns'])
         last_average_return = np.mean(self.episode_rewards)
         for train_itr in range(self.n_train_steps):
-            if (self.replay_buffer.n_transitions_stored >=
-                    self.min_buffer_size):
+            # yapf: disable
```
Comment (ryanjulian) [inline src/garage/tf/algos/dqn.py]: it seems like YAPF is misbehaving for comparison or logical operators?

File: src/garage/tf/algos/dqn.py [hunk 4]
```diff
@@ -80,22 +80,21 @@ def __init__(self,
...
-            n_epoch_cycles=n_epoch_cycles,
-            buffer_batch_size=buffer_batch_size,
-            rollout_batch_size=rollout_batch_size,
-            replay_buffer=replay_buffer,
-            max_path_length=max_path_length,
-            discount=discount,
-            reward_scale=reward_scale,
-            input_include_goal=input_include_goal,
-            smooth_return=smooth_return)
+        super(DQN, self).__init__(env_spec=env_spec,
```
Comment (zequnyu) [inline src/garage/tf/algos/dqn.py]: `super().__init__()`

File: src/garage/tf/algos/dqn.py (hunk 3 above)
Comment (gitanshu) [reply src/garage/tf/algos/dqn.py]: It works fine at some places for logical operators like https://github.com/rlworkgroup/garage/pull/773/files#diff-eec0e5bc5f0646f0458d535037a7bcf0R363 & https://github.com/rlworkgroup/garage/pull/773/files#diff-eec0e5bc5f0646f0458d535037a7bcf0R363
but goes crazy at others. And yapf doesn't have a flag that covers comparison operators at all

Comment (gitanshu) [reply src/garage/tf/algos/dqn.py]: It also messed up around some arithmetic operators sometimes like if the an arithmetic expression is passed to a function call such as [here](https://github.com/rlworkgroup/garage/pull/773/files#diff-bb9dfa32db8aec7c259bc87d036131adR53)

Comment (ryanjulian) [reply src/garage/tf/algos/dqn.py]: i think comparison operators are considered logical operators -- but yes every time i've tried this, YAPF has had terrible W504 support.

you may also want to try `split_before_bitwise_operator`

you can increase `split_penalty_arithmetic_operator`, `split_penalty_bitwise_operator` and `split_penalty_bitwise_operator` if that doesn't work.  the defaults are here: https://github.com/google/yapf/blob/45cf63a4538a2d5790db5182b84aca8c1b3bc491/yapf/yapflib/style.py#L341 -- just try doubling them until it behaves correctly.

Comment (gitanshu) [reply src/garage/tf/algos/dqn.py]: [yapf source](https://github.com/google/yapf/blob/27fe4ad04fd61979ae8b734d13a569a29a3b79bd/yapf/yapflib/unwrapped_line.py#L520) has only 'and' and 'or' as logical operators. I don't see comparison operators mentioned at all.

Increasing the penalties helps 1 or 2 cases of logical operators in terms of line break but still not the best for readability. It largely does no good for arithmetic operators, and instead increasing it too much for arithmetic operators causes unexpected changes. 

TLDR: even with all these knobs and high penalties, yapf doesn't produce expected output

**Comment by ryanjulian** [comment]
can you also address #758 in this PR?

**Comment by gitanshu** [comment]
I tried switching to the default pre-commit flake8 hook [as described here](https://flake8.pycqa.org/en/latest/user/using-hooks.html) but it runs into the [pydocstyle version issue](https://gitlab.com/pycqa/flake8-docstrings/issues/36). I am guessing pre-commit runs the cloned repos in it's own environment which is why the version of pydocstyle we pin in setup.py doesn't get picked up here

**Comment by ryanjulian** [comment]
@gitanshu to fix the pydocstyle issue, look at how we run pylint in .pre-commit-config.yaml

**Comment by gitanshu** [comment]
> @gitanshu to fix the pydocstyle issue, look at how we run pylint in .pre-commit-config.yaml

Yeah I realised that after my last comment here. Addressed it in https://github.com/rlworkgroup/garage/pull/773/commits/19a0594d86cf129f6abbabb10e8c83fada4955f1

**Comment by ryanjulian** [comment]
What triggered all the changes to hanging indentation? Was it your tweaks to the YAPF config?

I am starting to think the simple solution for now is to disable W504 in the flake8 config. I don't think our tools are ready to support it yet.

**Comment by gitanshu** [comment]
@ryanjulian it could be the version bump from 0.25 to 0.27. There might be a knob to disable it in yapf. I could look into it. But I think there's still gonna be a number of yapf disables in future code too.

As for disabling W504, I guess we could just leave/close this PR and revisit it later when tools support it better.

**Comment by ryanjulian** [comment]
i think removing enforcement of W504 (and any custom YAPF stuff you had to add) + check_flake8.sh would be a net win for the project, because it would simplify the CI pipeline and unify how code is checked for style (the rules in check_flake8 are a little complicated and it tends to break).

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -9,9 +9,11 @@ repos:
         stages: [commit-msg]
     -   id: check-flake8
         name: Check Flake8
-        language: script
-        entry: scripts/travisci/check_flake8.sh
+        language: system
```
Comment (ryanjulian) [inline .pre-commit-config.yaml]: is it not sufficient to pin the flake8 version both here and in setup.py?

Comment (gitanshu) [reply .pre-commit-config.yaml]: Do you mean without mentioning a language? It's a mandatory key for pre-commit and `system` seems like the most relevant value. `python` could work but it tries to install flake-docstrings in a virtual env and fails on the pydocstyle issue.

Comment (ryanjulian) [reply .pre-commit-config.yaml]: i mean not using system at all (just the default pre-commit rule), but pinning the version to match ours.

Comment (gitanshu) [reply .pre-commit-config.yaml]: We don't pin flake8 version yet, but we could do something like this

```

-   repo: https://gitlab.com/pycqa/flake8
    rev: '3.7.8'
    hooks:
    -   id: flake8
        stages: [commit, push]
        additional_dependencies: ['flake8-import-order', 'flake8-docstrings==1.3.0', 'pydocstyle<4.0.0']
```

Is this what you meant?

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -9,9 +9,11 @@ repos:
         stages: [commit-msg]
     -   id: check-flake8
         name: Check Flake8
-        language: script
-        entry: scripts/travisci/check_flake8.sh
+        entry: flake8
```
Comment (ryanjulian) [inline .pre-commit-config.yaml]: is there a reason we need to add our own `system` entry, instead of just using the default pre-commit flake8 rule.

Comment (ryanjulian) [reply .pre-commit-config.yaml]: i.e. here https://github.com/rlworkgroup/garage/pull/773/files#diff-82c3b36709926de3a1b27bea0d8258ccL70

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -9,9 +9,11 @@ repos:
         stages: [commit-msg]
```
Comment (ryanjulian) [inline .pre-commit-config.yaml]: please take a moment to remove the commented-out entries in this file. they are no longer necessary

Comment (ryanjulian) [inline .pre-commit-config.yaml]: the `stages` key is now deprecated, and defaults to our config, so you can remove all of these `stages` keys.

Comment (ryanjulian) [reply .pre-commit-config.yaml]: (except for the commit-msg one, which is not default)

Comment (ryanjulian) [reply .pre-commit-config.yaml]: okay i am now less sure of this change, so please test it, but i think it would help clean up this file a bit.

Comment (ryanjulian) [reply .pre-commit-config.yaml]: ok i confirmed the behavior -- it is save to remove the stages keys.

File: setup.py [hunk 8]
```diff
@@ -70,7 +70,7 @@
     'recommonmark',
     'sphinx',
     'sphinx_rtd_theme',
-    'yapf',
+    'yapf==0.25.0',
```
Comment (gitanshu) [inline setup.py]: @ryanjulian If we are setting yapf to 0.28.0 in pre-commit, this should also be 0.28.0, right?

Comment (ryanjulian) [reply setup.py]: yes -- we should strive to have them all match.

**Comment by gitanshu** [comment]
Moved to #844
